### PR TITLE
core(persistence): struct storage layout without alignment gaps

### PR DESCRIPTION
**Pull request by alalek**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/636

relates #14977

Align with OpenCV 4.x rewritten implementation.
Rework tests.

```
allow_multiple_commits=1
```

### CODE REVIEW

File: modules/core/src/persistence_base64.cpp [hunk 1]
```diff
@@ -803,9 +811,28 @@ class BinaryToCvSeqConvertor
...
+                    static bool skip_message = cv::utils::getConfigurationParameterBool("OPENCV_PERSISTENCE_SKIP_PACKED_STRUCT_WARNING",
+#ifdef _DEBUG
+                            false
+#else
+                            true
+#endif
+                    );
+                    if (!skip_message)
+                    {
+                        CV_LOG_WARNING(NULL, "Binary converter: struct storage layout has been changed in OpenCV 3.4.7. Alignment gaps has been removed from the storage containers. "
```
Comment (mshabunin) [inline modules/core/src/persistence_base64.cpp]: Can we use `CV_LOG_DEBUG` and avoid introduction of new environment configuration parameter?

File: modules/core/include/opencv2/core/persistence.hpp [hunk 2]
```diff
@@ -597,8 +597,8 @@ class CV_EXPORTS_W_SIMPLE FileNode
     Usually it is more convenient to use operator `>>` instead of this method.
     @param fmt Specification of each array element. See @ref format_spec "format specification"
     @param vec Pointer to the destination array.
-    @param len Number of elements to read. If it is greater than number of remaining elements then all
-    of them will be read.
+    @param len Number of bytes to read (buffer size limit). If it is greater than number of
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/persistence.hpp]:
```suggestion
    @param len Number of bytes to read (buffer size limit). If it is greater than size of
```

File: modules/core/include/opencv2/core/persistence.hpp [hunk 3]
```diff
@@ -668,11 +668,12 @@ class CV_EXPORTS FileNodeIterator
     Usually it is more convenient to use operator `>>` instead of this method.
     @param fmt Specification of each array element. See @ref format_spec "format specification"
     @param vec Pointer to the destination array.
-    @param maxCount Number of elements to read. If it is greater than number of remaining elements then
-    all of them will be read.
+    @param len Number of bytes to read (buffer size limit). If it is greater than number of
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/persistence.hpp]:
```suggestion
    @param len Number of bytes to read (buffer size limit). If it is greater than size of
```

File: modules/core/src/persistence_base64.cpp (hunk 1 above)
Comment (alalek) [reply modules/core/src/persistence_base64.cpp]: `DEBUG` messages are not compiled into release builds by default

**Pull request merged**